### PR TITLE
sink(ticdc): optimize buffer sink flush from O(N^2) to O(N)

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve? 

Optimize buffer sink flush from O(N^2) to O(N).

O(N^2): It flushes all tables for every `flushEvent`, and each table will periodically generate `flushEvent`s.
O(N): It flushes tables according to `flushEvent.tableID`.

Close https://github.com/pingcap/ticdc/issues/3900

Benchmark:

```sh
$ go test -benchmem -run='^$' -bench '^BenchmarkRun$' github.com/pingcap/ticdc/cdc/sink
goos: linux
goarch: amd64
pkg: github.com/pingcap/ticdc/cdc/sink
cpu: Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz
BenchmarkRun/1_table(s)-40                745132              1566 ns/op              16 B/op          1 allocs/op
BenchmarkRun/4_table(s)-40                477426              2459 ns/op              64 B/op          4 allocs/op
BenchmarkRun/16_table(s)-40               171915              6228 ns/op             256 B/op         16 allocs/op
BenchmarkRun/64_table(s)-40                57037             22452 ns/op            1024 B/op         64 allocs/op
BenchmarkRun/256_table(s)-40               12612             89534 ns/op            4096 B/op        256 allocs/op
BenchmarkRun/1024_table(s)-40               3826            325067 ns/op           16384 B/op       1024 allocs/op
BenchmarkRun/4096_table(s)-40                854           1312032 ns/op           65536 B/op       4096 allocs/op
BenchmarkRun/16384_table(s)-40               208           5243129 ns/op          262145 B/op      16384 allocs/op
BenchmarkRun/65536_table(s)-40                55          19329650 ns/op         1048577 B/op      65536 allocs/op

$ benchstat bin/master.log bin/opt-sink.log
name                   old time/op    new time/op    delta
Run/1_table(s)-40        1.04µs ± 0%    1.66µs ±14%     ~     (p=0.333 n=1+5)
Run/4_table(s)-40        8.75µs ± 0%    2.44µs ± 3%     ~     (p=0.333 n=1+5)
Run/16_table(s)-40        103µs ± 0%       6µs ± 2%     ~     (p=0.333 n=1+5)
Run/64_table(s)-40       1.48ms ± 0%    0.02ms ± 3%     ~     (p=0.333 n=1+5)
Run/256_table(s)-40      22.8ms ± 0%     0.1ms ± 7%     ~     (p=0.333 n=1+5)
Run/1024_table(s)-40      368ms ± 0%       0ms ± 4%     ~     (p=0.333 n=1+5)
Run/4096_table(s)-40      6.01s ± 0%     0.00s ± 3%     ~     (p=0.333 n=1+5)
Run/16384_table(s)-40     98.0s ± 0%      0.0s ± 5%     ~     (p=0.333 n=1+5)

name                   old alloc/op   new alloc/op   delta
Run/1_table(s)-40         24.0B ± 0%     16.0B ± 0%     ~     (p=1.667 n=1+5)
Run/4_table(s)-40          192B ± 0%       64B ± 0%     ~     (p=1.667 n=1+5)
Run/16_table(s)-40       2.30kB ± 0%    0.26kB ± 0%     ~     (p=1.667 n=1+5)
Run/64_table(s)-40       33.8kB ± 0%     1.0kB ± 0%     ~     (p=1.667 n=1+5)
Run/256_table(s)-40       528kB ± 0%       4kB ± 0%     ~     (p=1.667 n=1+5)
Run/1024_table(s)-40     8.40MB ± 0%    0.02MB ± 0%     ~     (p=1.667 n=1+5)
Run/4096_table(s)-40      134MB ± 0%       0MB ± 0%     ~     (p=1.667 n=1+5)
Run/16384_table(s)-40    2.15GB ± 0%    0.00GB ± 0%  -99.99%  (p=0.000 n=1+5)

name                   old allocs/op  new allocs/op  delta
Run/1_table(s)-40          2.00 ± 0%      1.00 ± 0%     ~     (p=1.667 n=1+5)
Run/4_table(s)-40          20.0 ± 0%       4.0 ± 0%     ~     (p=1.667 n=1+5)
Run/16_table(s)-40          272 ± 0%        16 ± 0%     ~     (p=1.667 n=1+5)
Run/64_table(s)-40        4.16k ± 0%     0.06k ± 0%     ~     (p=1.667 n=1+5)
Run/256_table(s)-40       65.8k ± 0%      0.3k ± 0%     ~     (p=1.667 n=1+5)
Run/1024_table(s)-40      1.05M ± 0%     0.00M ± 0%     ~     (p=1.667 n=1+5)
Run/4096_table(s)-40      16.8M ± 0%      0.0M ± 0%     ~     (p=1.667 n=1+5)
Run/16384_table(s)-40      268M ± 0%        0M ± 0%     ~     (p=1.667 n=1+5)
```

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Reduce checkpoint lag when capturing many tables.
```

### CODE REVIEW

**Review by asddongmen — APPROVED** [review]

**Comment by sdojjy** [comment]
Can you add some description about how it works?

**Comment by overvenus** [comment]
> Can you add some description about how it works?

Updated, see description.

File: cdc/sink/buffer_sink.go [hunk 1]
```diff
@@ -38,99 +41,126 @@ type bufferSink struct {
...
+				break RecvBatch
 			}
+		}
+	}
+
+	b.bufferMu.Lock()
+	startEmit := time.Now()
+	// find all rows before resolvedTs and emit to backend sink
+	for i := 0; i < batchSize; i++ {
+		tableID, resolvedTs := batch[i].tableID, batch[i].resolvedTs
```
Comment (sdojjy) [inline cdc/sink/buffer_sink.go:122]: if there are many flush events in `batch` for one table, does it help to improve the performance to find the max resovedTs per table first,  do not need to search and flush multi times?

File: cdc/sink/manager.go [hunk 2]
```diff
@@ -94,36 +86,8 @@ func (m *Manager) Close(ctx context.Context) error {
...
-}
-
-func (m *Manager) flushBackendSink(ctx context.Context, tableID model.TableID) (model.Ts, error) {
-	// NOTICE: Because all table sinks will try to flush backend sink,
-	// which will cause a lot of lock contention and blocking in high concurrency cases.
-	// So here we use flushing as a lightweight lock to improve the lock competition problem.
-	if !atomic.CompareAndSwapInt64(&m.flushing, 0, 1) {
-		return m.getCheckpointTs(tableID), nil
-	}
-	m.flushMu.Lock()
```
Comment (sdojjy) [inline cdc/sink/manager.go:120]: FYI,  mq sink is not thread-safe

Comment (overvenus) [reply cdc/sink/manager.go:120]: Manager flushes bufferSink concurrently, and bufferSink itself is thread-safe.

**Review by amyangfei — APPROVED** [review]

**Review by zhaoxinyu — APPROVED** [review]

**Pull request merged**